pkg/dynamicaccess/controller.go

```diff
@@ -1,54 +1,210 @@
-	kvs, err := c.history.Lookup(timestamp)
+var _ Controller = (*controller)(nil)
+
+func (c *controller) DownloadHandler(
+	ctx context.Context,
+	timestamp int64,
+	encryptedRef swarm.Address,
+	publisher *ecdsa.PublicKey,
+	historyRootHash swarm.Address,
+) (swarm.Address, error) {
+	ls := loadsave.New(c.getter, c.putter, requestPipelineFactory(ctx, c.putter, false, redundancy.NONE))
+	history, err := NewHistory(ls, &historyRootHash)
 	if err != nil {
-		return swarm.EmptyAddress, err
+		return swarm.ZeroAddress, err
 	}
-	addr, err := c.accessLogic.DecryptRef(kvs, enryptedRef, publisher)
-	return addr, err
+
+	kvsRef, err := history.Lookup(ctx, timestamp)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	kvs := kvs.New(ls, kvsRef)
+	return c.accessLogic.DecryptRef(ctx, kvs, encryptedRef, publisher)
 }
 
-func (c *defaultController) UploadHandler(ref swarm.Address, publisher *ecdsa.PublicKey) (swarm.Address, error) {
-	kvs, err := c.history.Lookup(0)
+// TODO: review return params: how to get back history ref ?
+func (c *controller) UploadHandler(
+	ctx context.Context,
+	refrefence swarm.Address,
+	publisher *ecdsa.PublicKey,
+	historyRootHash *swarm.Address,
+) (swarm.Address, swarm.Address, swarm.Address, error) {
+	ls := loadsave.New(c.getter, c.putter, requestPipelineFactory(ctx, c.putter, false, redundancy.NONE))
+	history, err := NewHistory(ls, historyRootHash)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	now := time.Now().Unix()
+	kvsRef, err := history.Lookup(ctx, now)
 	if err != nil {
-		return swarm.EmptyAddress, err
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
 	}
-	if kvs == nil {
-		// new feed
-		// TODO: putter session to create kvs
-		kvs = kvsmock.New()
-		_, err = c.granteeManager.Publish(kvs, publisher)
+	kvs := kvs.New(ls, kvsRef)
+	historyRef := *historyRootHash
+	if kvsRef.Equal(swarm.ZeroAddress) {
+		err = c.accessLogic.AddPublisher(ctx, kvs, publisher)
 		if err != nil {
-			return swarm.EmptyAddress, err
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		kvsRef, err = kvs.Save(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		err = history.Add(ctx, kvsRef, &now)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		historyRef, err = history.Store(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+	encryptedRef, err := c.accessLogic.EncryptRef(ctx, kvs, publisher, refrefence)
+	return kvsRef, historyRef, encryptedRef, err
+}
+
+func NewController(ctx context.Context, accessLogic ActLogic, getter storage.Getter, putter storage.Putter) Controller {
+	return &controller{
+		granteeList: nil, //NewGranteeList(ls, ps, swarm.EmptyAddress),
```

Comment:
NewGranteeList commented line is still needed or relevant?

Reply:
Gratnee related functions will be handled in a later PR.